include/deal.II/lac/solver_gmres.h

```diff
@@ -710,81 +723,319 @@ SolverGMRES<VectorType>::givens_rotation(Vector<double> &h,
+      const LinearAlgebra::distributed::Vector<Number, MemorySpace::Host> &vv,
+      const internal::SolverGMRESImplementation::TmpVectors<
+        LinearAlgebra::distributed::Vector<Number, MemorySpace::Host>>
+        &             orthogonal_vectors,
+      Vector<double> &h)
     {
-      // Orthogonalization
-      double htmp = vv * orthogonal_vectors[0];
-      h(0) += htmp;
+      AlignedVector<VectorizedArray<double>> hs(dim,
+                                                VectorizedArray<double>(0.0));
+
+      unsigned int j = 0;
+      for (unsigned int c = 0;
+           j < vv.locally_owned_size() / VectorizedArray<double>::size();
+           ++c, j += VectorizedArray<double>::size())
+        for (unsigned int i = 0; i < dim; ++i)
+          DEAL_II_OPENMP_SIMD_PRAGMA
+      for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+        hs[i][v] +=
+          orthogonal_vectors[i].local_element(j + v) * vv.local_element(j + v);
+
+      for (unsigned int i = 0; i < dim; ++i)
+        for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+          h(i) += hs[i][v];
+
+      for (; j < vv.locally_owned_size(); ++j)
+        for (unsigned int i = 0; i < dim; ++i)
+          h(i) += orthogonal_vectors[i].local_element(j) * vv.local_element(j);
+
+      Utilities::MPI::sum(h, MPI_COMM_WORLD, h);
+    }
+
+
+
+    template <class VectorType>
+    double
+    substract_and_norm(
+      const unsigned int dim,
+      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+        &                   orthogonal_vectors,
+      const Vector<double> &h,
+      VectorType &          vv)
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        vv.add(-h(i), orthogonal_vectors[i]);
+
+      return vv.l2_norm();
+    }
+
+
+
+    template <class Number>
+    double
+    substract_and_norm(
+      const unsigned int dim,
+      const internal::SolverGMRESImplementation::TmpVectors<
+        LinearAlgebra::distributed::Vector<Number, MemorySpace::Host>>
+        &                   orthogonal_vectors,
+      const Vector<double> &h,
+      LinearAlgebra::distributed::Vector<Number, MemorySpace::Host> &vv)
+    {
+      double                  norm_vv_temp            = 0.0;
+      VectorizedArray<double> norm_vv_temp_vectorized = 0.0;
+
+      unsigned int j = 0;
+      for (unsigned int c = 0;
+           j < vv.locally_owned_size() / VectorizedArray<double>::size();
+           ++c, j += VectorizedArray<double>::size())
+        {
+          VectorizedArray<double> temp;
+
+          DEAL_II_OPENMP_SIMD_PRAGMA
+          for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+            temp[v] = vv.local_element(j + v);
+
+          for (unsigned int i = 0; i < dim; ++i)
+            DEAL_II_OPENMP_SIMD_PRAGMA
+          for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+            temp[v] -= h(i) * orthogonal_vectors[i].local_element(j + v);
```

Comment:
Not relevant for the big sizes considered now, but this could be substantially sped up for the in-cache case when expressing this loop such that the factor `h(i)` could be re-used for several loop iterations (but not too many to keep `temp` in registers).